linfa-ensemble/src/random_forest/algorithm.rs

```diff
@@ -0,0 +1,122 @@
+use linfa_trees::DecisionTree;
+use crate::random_forest::hyperparameters::RandomForestParams;
+use ndarray::{Array1, Array2, ArrayBase, Data, Ix1, Ix2};
+use ndarray_rand::rand_distr::Uniform;
+use ndarray::Axis;
+use ndarray::Array;
+use ndarray_rand::RandomExt;
+use std::collections::HashMap;
+
+
+pub struct RandomForest {
+    pub hyperparameters: RandomForestParams,
+    pub trees: Vec<DecisionTree>,
+}
+
+impl RandomForest {
+    pub fn fit(
+        hyperparameters: RandomForestParams,
+        x: &ArrayBase<impl Data<Elem = f64>, Ix2>,
+        y: &ArrayBase<impl Data<Elem = u64>, Ix1>,
+    ) -> Self {
+
+    let n_estimators = hyperparameters.n_estimators;
+    let mut trees: Vec<DecisionTree> = Vec::with_capacity(n_estimators);
+    let single_tree_params = hyperparameters.tree_hyperparameters;
+
+    //TODO check bootstrap
+    let _bootstrap = hyperparameters.bootstrap;
+
+    for _ in 0..n_estimators {
+        // Bagging here
+        let rnd_idx = Array::random((1, x.nrows()), Uniform::new(0, x.nrows())).into_raw_vec();
```

Comment:
perhaps it would make sense to let the user choose a sample size smaller than the original one

Reply:
why?
the official implementation of RF with bagging uses the same sample size with replacement (to encourage diversity in the sample to fit the single tree).
If such sample is too small, there is a risk of fitting identical trees.

Reply:
not necessary and rather an edge case, I was just thinking about a use-case where you have a very large dataset and want to speed up the inference

Reply:
Got it. Then I can define an `u64` to indicate the number of rows of data to fit trees against.